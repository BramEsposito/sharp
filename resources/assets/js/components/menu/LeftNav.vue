<template>
    <div>
        <nav v-show="ready" class="SharpLeftNav SharpLeftNav--collapseable" :class="`SharpLeftNav--${state}`"
             role="navigation" aria-label="Menu Sharp" @click="collapsed && (collapsed=false)">
            <div class="SharpLeftNav__top-icon">
                <i class="fa" :class="currentIcon"></i>
            </div>
            <slot></slot>
            <div class="SharpLeftNav__collapse" @click.stop="collapsed = !collapsed">
                <a class="SharpLeftNav__collapse-link" href="#" @click.prevent>
                    <svg class="SharpLeftNav__collapse-arrow" width="8" height="12" viewBox="0 0 8 12" fill-rule="evenodd">
                        <path d="M7.5 10.6L2.8 6l4.7-4.6L6.1 0 0 6l6.1 6z"></path>
                    </svg>
                </a>
            </div>
        </nav>
    </div>
</template>

<script>
    import { Responsive } from '../../mixins';

    export default {
        name: 'SharpLeftNav',

        mixins: [Responsive('lg')],

        props: {
            categories: Array,
            current: String
        },
        data() {
            return {
                collapsed: null,
                state: 'expanded',
                ready: false
            }
        },
        watch: {
            collapsed: {
                immediate: true,
                handler(val, oldVal) {
                    if(oldVal === null) {
                        return this.updateState();
                    }
                    this.$root.$emit('setClass', 'leftNav--collapsed', this.collapsed);
                    // apply transition
                    this.state = val ? 'collapsing' : 'expanding';
                    setTimeout(this.updateState, 250);
                }
            }
        },
        computed: {
            allEntities() {
                return this.categories.reduce((res,category)=>[...res,...category.entities],[]);
            },
            currentIcon() {
                return this.current === 'dashboard'
                    ? 'fa-dashboard'
                    : (this.allEntities.find(e=>e.key===this.current)||{}).icon;
            }
        },
        methods: {
            updateState() {
                this.state = this.collapsed ? 'collapsed' : 'expanded';
            }
        },
        mounted() {
            this.collapsed = this.isViewportSmall;
            this.$nextTick(_=>this.ready=true);
        }
    }
</script>