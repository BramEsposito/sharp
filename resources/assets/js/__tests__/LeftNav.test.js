import { mount } from '@vue/test-utils';
import LeftNav from '../components/menu/LeftNav.vue';
import Vue from 'vue';
jest.useFakeTimers();



describe('left-nav', ()=>{
    let wrapper;
    function createWrapper(options) {
        let vm = mount(LeftNav, {
            slots: {
                default: '<div>NAV CONTENT</div>'
            },
            propsData: {
                categories: [{ entities:[] }]
            },
            created() {
                jest.spyOn(this,'updateState');
            },
            ...options
        });
        vm.setData({ ready:true });
        jest.runAllTimers();
        jest.runAllTicks();
        return vm;
    }
    beforeEach(()=>{

    });

    test('can mount LeftNav', ()=>{
        expect(createWrapper().html()).toMatchSnapshot();
    });
    test('can mount "not ready" LeftNav', ()=>{
        const wrapper = createWrapper();
        wrapper.setData({ ready:false });
        expect(wrapper.html()).toMatchSnapshot();
    });
    test('can mount "with state" LeftNav', ()=>{
        const wrapper = createWrapper();
        wrapper.setData({ state:'STATE' });
        expect(wrapper.html()).toMatchSnapshot();
    });
    test('can mount "with current icon" LeftNav', ()=>{
        let wrapper = createWrapper({ computed:{ currentIcon:()=>'CURRENT_ICON' }});
        expect(wrapper.html()).toMatchSnapshot();
    });


    describe('watch collapsed', ()=>{
        test('is immediate', ()=>{
            expect(createWrapper().vm.$options.watch.collapsed.immediate).toBe(true);
        });

        test('updating state for intermediate animations and set root class', async ()=>{
            const wrapper = createWrapper({ sync:false });
            const collapsedClass = 'leftNav--collapsed';

            await Vue.nextTick();

            expect(wrapper.emitted()['setClass']).toHaveLength(1);
            expect(wrapper.vm.updateState).toHaveBeenCalledTimes(1);

            wrapper.setData({ collapsed: true });

            await Vue.nextTick();

            expect(wrapper.emitted()['setClass'][1]).toEqual([collapsedClass, true]);
            expect(setTimeout).toHaveBeenCalledWith(wrapper.vm.updateState, 250); // update state called at the end of the animation
            expect(wrapper.vm.state).toBe('collapsing');

            wrapper.setData({ collapsed: false });

            await Vue.nextTick();

            expect(wrapper.emitted()['setClass'][2]).toEqual([collapsedClass, false]);
            expect(wrapper.vm.state).toBe('expanding');

            jest.runOnlyPendingTimers();
            expect(wrapper.vm.updateState).toHaveBeenCalledTimes(3);
        });
    });

    test('allEntities', ()=>{
        const wrapper = createWrapper();
        wrapper.setProps({
            categories: [
                { entities:[1] },
                { entities:[2] },
            ]
        });
        expect(wrapper.vm.allEntities).toEqual([1,2]);
    });

    test('currentIcon', ()=>{
        let wrapper = createWrapper();
        wrapper.setProps({ current:'dashboard' });
        expect(wrapper.vm.currentIcon).toBe('fa-dashboard');

        wrapper = createWrapper({
            computed: {
                allEntities:()=>[
                    { key:1, icon:'firstIcon' },
                    { key:2, icon:'secondIcon' }
                ]
            }
        });

        wrapper.setProps({ current:1 });
        expect(wrapper.vm.currentIcon).toBe('firstIcon');
        wrapper.setProps({ current:2 });
        expect(wrapper.vm.currentIcon).toBe('secondIcon');
    });

    test('updateState', ()=>{
        const wrapper = createWrapper();
        wrapper.setData({ collapsed: true });
        wrapper.vm.updateState();
        expect(wrapper.vm.state).toBe('collapsed');

        wrapper.setData({ collapsed: false });
        wrapper.vm.updateState();
        expect(wrapper.vm.state).toBe('expanded');
    });

    test('mounted hook', ()=>{
        const wrapper = mount({
            render:()=>null,
            data:()=>({
                isViewportSmall: false,
                collapsed: null,
                ready: false
            })
        });
        LeftNav.mounted.call(wrapper.vm);
        expect(wrapper.vm.collapsed).toBe(false);

        wrapper.setData({ isViewportSmall: true });
        LeftNav.mounted.call(wrapper.vm);
        expect(wrapper.vm.collapsed).toBe(true);
    });
});