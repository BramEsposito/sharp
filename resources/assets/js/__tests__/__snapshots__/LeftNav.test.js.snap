// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`left-nav can mount "not ready" LeftNav 1`] = `
<div>
	<nav role="navigation" aria-label="Menu Sharp" class="SharpLeftNav SharpLeftNav--collapseable SharpLeftNav--expanded" style="display: none;">
		<div class="SharpLeftNav__top-icon">
			<i class="fa"></i>
		</div>
		<div>NAV CONTENT</div>
		<div class="SharpLeftNav__collapse">
			<a href="#" class="SharpLeftNav__collapse-link">
				<svg width="8" height="12" viewBox="0 0 8 12" fill-rule="evenodd" class="SharpLeftNav__collapse-arrow">
					<path d="M7.5 10.6L2.8 6l4.7-4.6L6.1 0 0 6l6.1 6z"></path>
				</svg>
			</a>
		</div>
	</nav>
</div>
`;

exports[`left-nav can mount "with current icon" LeftNav 1`] = `
<div>
	<nav role="navigation" aria-label="Menu Sharp" class="SharpLeftNav SharpLeftNav--collapseable SharpLeftNav--expanded" style="">
		<div class="SharpLeftNav__top-icon">
			<i class="fa CURRENT_ICON"></i>
		</div>
		<div>NAV CONTENT</div>
		<div class="SharpLeftNav__collapse">
			<a href="#" class="SharpLeftNav__collapse-link">
				<svg width="8" height="12" viewBox="0 0 8 12" fill-rule="evenodd" class="SharpLeftNav__collapse-arrow">
					<path d="M7.5 10.6L2.8 6l4.7-4.6L6.1 0 0 6l6.1 6z"></path>
				</svg>
			</a>
		</div>
	</nav>
</div>
`;

exports[`left-nav can mount "with state" LeftNav 1`] = `
<div>
	<nav role="navigation" aria-label="Menu Sharp" class="SharpLeftNav SharpLeftNav--collapseable SharpLeftNav--STATE" style="">
		<div class="SharpLeftNav__top-icon">
			<i class="fa"></i>
		</div>
		<div>NAV CONTENT</div>
		<div class="SharpLeftNav__collapse">
			<a href="#" class="SharpLeftNav__collapse-link">
				<svg width="8" height="12" viewBox="0 0 8 12" fill-rule="evenodd" class="SharpLeftNav__collapse-arrow">
					<path d="M7.5 10.6L2.8 6l4.7-4.6L6.1 0 0 6l6.1 6z"></path>
				</svg>
			</a>
		</div>
	</nav>
</div>
`;

exports[`left-nav can mount LeftNav 1`] = `
<div>
	<nav role="navigation" aria-label="Menu Sharp" class="SharpLeftNav SharpLeftNav--collapseable SharpLeftNav--expanded" style="">
		<div class="SharpLeftNav__top-icon">
			<i class="fa"></i>
		</div>
		<div>NAV CONTENT</div>
		<div class="SharpLeftNav__collapse">
			<a href="#" class="SharpLeftNav__collapse-link">
				<svg width="8" height="12" viewBox="0 0 8 12" fill-rule="evenodd" class="SharpLeftNav__collapse-arrow">
					<path d="M7.5 10.6L2.8 6l4.7-4.6L6.1 0 0 6l6.1 6z"></path>
				</svg>
			</a>
		</div>
	</nav>
</div>
`;
